crates/gitbutler-branch-actions/src/virtual.rs

```diff
@@ -860,6 +848,104 @@ pub fn commit(
     Ok(commit_oid)
 }
 
+fn join_output<'a>(stdout: &'a str, stderr: &'a str) -> Cow<'a, str> {
+    let stdout = stdout.trim();
+    if stdout.is_empty() {
+        stderr.trim().into()
+    } else {
+        stdout.into()
+    }
+}
+
+fn run_message_hook(ctx: &CommandContext, mut message: String) -> Result<String> {
+    let hook_result = git2_hooks::hooks_commit_msg(ctx.repo(), Some(&["../.husky"]), &mut message)
+        .context("failed to run hook")
+        .context(Code::CommitHookFailed)?;
+
+    if let HookResult::RunNotSuccessful { stdout, stderr, .. } = &hook_result {
+        return Err(
+            anyhow!("commit-msg hook rejected: {}", join_output(stdout, stderr))
+                .context(Code::CommitHookFailed),
+        );
+    }
+    Ok(message)
+}
+
+fn run_pre_commit_hook(ctx: &CommandContext) -> Result<()> {
+    let hook_result = git2_hooks::hooks_pre_commit(ctx.repo(), Some(&["../.husky"]))
+        .context("failed to run hook")
+        .context(Code::CommitHookFailed)?;
+
+    if let HookResult::RunNotSuccessful { stdout, stderr, .. } = &hook_result {
+        return Err(
+            anyhow!("commit hook rejected: {}", join_output(stdout, stderr))
+                .context(Code::CommitHookFailed),
+        );
+    }
+    Ok(())
+}
+
+pub(crate) fn unstage_all(ctx: &CommandContext) -> Result<()> {
+    let repo = ctx.repo();
+    // Get the HEAD commit (current commit)
+    let head_commit = repo.head()?.peel_to_commit()?;
+    // Reset the index to match the HEAD commit
+    repo.reset(head_commit.as_object(), ResetType::Mixed, None)?;
```

Comment:
At this time, GitButler doesn't differentiate between staged changes and unstaged ones, so resetting the index get get to the pre-staging state will be visually indistinguishable.

However, I think it would be more correct and more future-proof to go back to a previous state by writing the previous state held in memory, something like this.

```Rust
fn restore_index_on_error(repo, doit() -> Result) -> Result {
    let index = repo.index();
    let res = doit();
    if res.is_err() {
        index.write();
    }
    res
}
```

That way, changes to the index made by the user will be retained when restoring.

Reply:
Ah, nice! Let me do it that way.